src/ucp/wireup/wireup.c

```diff
@@ -1351,23 +1352,39 @@ ucs_status_t ucp_wireup_init_lanes(ucp_ep_h ep, unsigned ep_init_flags,
     if ((ep->cfg_index != UCP_WORKER_CFG_INDEX_NULL) &&
         /* reconfiguration is allowed for CM flow */
         !ucp_ep_has_cm_lane(ep)) {
-        /*
-         * TODO handle a case where we have to change lanes and reconfigure the ep:
-         *
-         * - if we already have uct ep connected to an address - move it to the new lane index
-         * - if we don't yet have connection to an address - create it
-         * - if an existing lane is not connected anymore - delete it (possibly)
-         * - if the configuration has changed - replay all pending operations on all lanes -
-         *   need that every pending callback would return, in case of failure, the number
-         *   of lane it wants to be queued on.
-         */
-        ucs_debug("cannot reconfigure ep %p from [%d] to [%d]", ep, ep->cfg_index,
-                  new_cfg_index);
-        ucp_wireup_print_config(worker, &ucp_ep_config(ep)->key, "old",
-                                NULL, cm_idx, UCS_LOG_LEVEL_ERROR);
-        ucp_wireup_print_config(worker, &key, "new", NULL,
-                                cm_idx, UCS_LOG_LEVEL_ERROR);
-        ucs_fatal("endpoint reconfiguration not supported yet");
+        if (can_reconfigure) {
+            for (lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
+                ucp_worker_discard_uct_ep(
+                        ep, ep->uct_eps[lane], UCT_FLUSH_FLAG_LOCAL,
+                        ucp_request_purge_enqueue_cb, &replay_pending_queue,
+                        (ucp_send_nbx_callback_t)ucs_empty_function, NULL);
+                ep->uct_eps[lane] = NULL;
+            }
+
+            ep->cfg_index = UCP_WORKER_CFG_INDEX_NULL;
```

Comment:
maybe reuse `ucp_ep_discard_lanes`? (then need to modify `ucp_ep_set_lanes_failed` to `ucp_ep_reset_lanes(..., <&ucp_failed_tl_ep|NULL>)`